src/datumaro/components/transformer.py

```diff
@@ -71,35 +73,87 @@ def __iter__(self):
 
 
 class ModelTransform(Transform):
+    """A transformation class for applying a model's inference to dataset items.
+
+    This class takes an dataset, a launcher, and other optional parameters
+    to transform the dataset item from the model outputs by the launcher.
+    It can process items using multiple processes if specified, making it suitable for
+    parallelized inference tasks.
+
+    Parameters:
+        extractor: The dataset extractor to obtain items from.
+        launcher: The launcher responsible for model inference.
+        batch_size: The batch size for processing items. Default is 1.
+        append_annotation: Whether to append inference annotations to existing annotations.
+            Default is False.
+        num_workers: The number of worker threads to use for parallel inference.
+            Set to 0 for single-process mode. Default is 0.
+    """
+
     def __init__(
         self,
         extractor: IDataset,
         launcher: Launcher,
         batch_size: int = 1,
         append_annotation: bool = False,
+        num_workers: int = 0,
     ):
         super().__init__(extractor)
         self._launcher = launcher
         self._batch_size = batch_size
         self._append_annotation = append_annotation
-
-    def __iter__(self) -> Generator[DatasetItem, None, None]:
-        for batch in take_by(self._extractor, self._batch_size):
-            inference = self._launcher.launch(
-                [item for item in batch if self._launcher.type_check(item)]
+        if not (isinstance(num_workers, int) and num_workers >= 0):
+            raise ValueError(
+                f"num_workers should be a non negative integer, but it is {num_workers}"
             )
-
-            for item in self._yield_item(batch, inference):
+        self._num_workers = num_workers
+
+    def __iter__(self) -> Iterator[DatasetItem]:
+        if self._num_workers == 0:
+            return self._iter_single_proc()
+        return self._iter_multi_procs()
+
+    def _iter_multi_procs(self):
+        with ThreadPool(processes=self._num_workers) as pool:
+
+            def _producer_gen():
+                for batch in take_by(self._extractor, self._batch_size):
+                    future = pool.apply_async(
+                        func=self._process_batch,
+                        args=(batch,),
+                    )
+                    yield future
+
+            with consumer_generator(producer_generator=_producer_gen) as consumer_gen:
+                for future in consumer_gen():
+                    for item in future.get():
+                        yield item
```

Comment:
can we adopt this kind of function for other transforms too through `consumer_generator`?

Reply:
Yes, but this is a helper to limit the size of requests at the same time (`queue_size: int = 100`). The actual one does parallelization is `ThreadPool`.
https://github.com/openvinotoolkit/datumaro/blob/ff99cec1bbee8ffb4d66d54b67d8741a40c9297c/src/datumaro/components/transformer.py#L116-L117

The many other transforms are based on the Python implementation. We need to use `multiprocessing.pool.Pool` not `ThreadPool` since Python has Global Interpreter Lock (GIL), so that one Python code line can be executed at the same time under GIL. The `multiprocessing.pool.Pool` spawns multiple processes, not threads, we can avoid it.